### ISSUE TITLE
path.js issues in Phonegap environment

### ISSUE THREAD

**Issue by kdonald** [mtrpcic/pathjs#33]
I have a web app (http://github.com/kdonald/graceworks) that is using pathjs client-side routing (hash fragments variety not push state). It is a multi-page mobile app. A back button is declared on each child page and when clicked we call window.history.back(). 

When testing in Chrome or another modern browser, everything works fine and in accordance with path.js documentation. For example, when you load the index page, the root route (#/) is fired. From there, when you activate links, other routes are fired as expected e.g (#/messages, #/messages/1, #/messages/1/parts/1)

However, when I wrap the web app's assets in a Phonegap/Cordova IOS project (native hybrid), I'm seeing strange behavior when invoking a back button on a child page of the root index page. Specifically, the back button on the child page, which invokes window.history.back(), does not work (nothing happens). Strangely, if I drill down 2-levels, for example, down from #/ to #/messages/ to #/messages/1 and back up to #/messages, the back button works fine. Once I try to back up to the root level (#/), the back button stops working completely.

So it seems something related to Path's root route implementation and the native IOS/Phonegap UiWebView is not playing together nicely. As an experiment, I tried Backbone's router as alternative. It seems to work fine with no issues, even backing up to the root index page.

You can see a running example of the strange behavior by cloning: http://github.com/kdonald/graceworks and then wrapping it in a Apache Cordova project as described here: http://docs.phonegap.com/en/1.8.1/guide_getting-started_ios_index.md.html#Getting%20Started%20with%20iOS.

You can see a Backbone example that works fully by cloning: https://github.com/ccoenraets/backbone-jquerymobile and wrapping it in a Apache Cordova project.

I also tried another routing framework (page.js) to see if it would work. Here's a issue tracking the results there: https://github.com/visionmedia/page.js/issues/18.

So in summary, I'm unable to get pathjs to work properly as a client-side routing framework in a native IOS environment with Phonegap on a UIWebView. It nearly works flawlessly, but going back to the first page recorded in history doesn't work and once that is attempted back behavior doesn't work again until the app is restarted.

As a side note, I noticed a difference between Backbone's router and the PathJS router: backbone has you register a root route as "" and does not append any fragment to the URL when it is fired. PathJS has you note a specific route e.g #/ that should be the root route, and when the root page is loaded the fragment portion of the URL is updated (e.g. #/ is appended).

cc @brianleroux

**Comment by kdonald** [comment]
I was able to resolve this with the following patch: https://github.com/kdonald/pathjs/commit/d20ca8c73cfb5bb11ca681d060e96ab502f96b2e.

I changed "root route" behavior to allow for registering an empty hash value ("") route, consistent with what Backbone allows. So when the root page loads and the hash value is "", this route executes automatically and the hash value remains un-changed. I tested this in a Phonegap environment and calling window.history.back() to go back to "" works as expected as well.

I'm still unclear why updating the hash via location.hash = Path.routes.root doesn't result in an entry getting put in history that you can go back to. Also, I'm not sure why it seems to break window.history.back() from that point on.

It's good to have a workaround at least. A weird issue though.

**Comment by brendo** [comment]
Thanks for posting a follow up comment. I have been looking for a way to map to 'nothing' as well and your patch allows me to do that. Thanks!

**Comment by derozn** [comment]
I found that by caching the location.href and setting this as the Page.base solved all issues i.e

let _href = window.location.href;

const Page = require('page');

Page({
	hashbang: true,
});

Page.base(_href);

Maybe not the best approach but works for me.
